File: src/main/java/list/RotateList.java
package list;

import structure.ListNode;

/**
 * https://leetcode-cn.com/problems/rotate-list/)
 */
public class RotateList {
    public static void main(String args[]){
        ListNode first = new ListNode(1);
        ListNode listNode = first;
        for (int i = 2; i < 6; i++) {
            listNode.next = new ListNode(i);
            listNode = listNode.next;
        }

        ListNode result = new RotateList().rotateRight(first, 2);
        while (result != null){
            System.out.println(result.val);
            result = result.next;
        }
    }

    public ListNode rotateRight(ListNode head, int k) {
        if (head == null)
            return null;

        int len = 1;
        ListNode point = head;
        while (point.next != null){
            point = point.next;
            len ++;
        }

        point.next = head;

        int moveNum = len - k % len - 1;
        for (int i = 0; i < moveNum; i++) {
            head = head.next;
        }

        ListNode tail = head;
        head = head.next;
        tail.next = null;
        return head;
    }

    public ListNode rotateRight1(ListNode head, int k) {
        if (head == null)
            return null;

        ListNode right = head;
        ListNode left = head;

        for (int i = 0; i < k; i++) {
            if (right.next != null)
                right = right.next;
            else
                right = head;
        }

        while (right.next != null){
            right = right.next;
            left = left.next;
        }

        right.next = head;
        head = left.next;
        left.next = null;

        return head;
    }
}

File: CppAlgorithms/list/RotateList.cpp
#include "../structure/ListNode.h"
using namespace std;

/**
 * https://leetcode-cn.com/problems/rotate-list/
 * 给你一个链表的头节点 head ，旋转链表，将链表每个节点向右移动 k 个位置。
 *
 * 输入：head = [1,2,3,4,5], k = 2
 * 输出：[4,5,1,2,3]
 *
 * 输入：head = [0,1,2], k = 4
 * 输出：[2,0,1]
 */
class RotateList {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL)
            return NULL;
        ListNode* right = head;
        ListNode* left = head;
        int size = 0;
        for (int i = 0; i < k; ++i) {
            ++size;
            if (right->next == NULL){
                int moveNum = k % size;
                right = head;
                for (int j = 0; j < moveNum; ++j) {
                    right = right->next;
                }
                break;
            }
            else
                right = right->next;
        }
        while (right->next != NULL){
            left = left->next;
            right = right->next;
        }
        right->next = head;
        head = left->next;
        left->next = NULL;
        return head;
    }
};

int main() {
    RotateList s;
    vector<int> v{0,1,2};
    ListNode listNode(v);
    s.rotateRight(&listNode, 4)->print();
}